### PR TITLE
Support `ExchangeType` for the response of annotated service

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

`ExchangeType` has been introduced with #3956. The changes in #3956
was huge, so I didn't implement `ExchangeType` for the response type of
an annotated service.

Modifications:

- Add `ResponseConverterFunction.isResponseStreaming()` to configure
  whether to enable response streaming.
  - The value is enabled by default because it abides by the original
    behavior.
  - The API uses `RequestHeaders` and a negotiated media type to
    determine response streaming.
- Override `.isResponseStreaming()` for all known implementions.
- Fix a bug where `ExchangeType` does not work for a `GrpcService`
  prefixed a path.
- Breaking) `HttpService.exchangeType()` now takes `RouteingContext` as
  a single parameter.
  - A negotiated media type was required.
- Add `RequestContext.exchangeType()` to get the `ExchangeType` of the current request.

Result:

- Annotated services perform better on unary responses.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/HttpService.java [hunk 1]
```diff
@@ -67,7 +68,8 @@ default HttpService decorate(DecoratingHttpServiceFunction function) {
      * <p>Note that an {@link HttpRequest} will be aggregated before serving the {@link HttpService} if
      * {@link ExchangeType#UNARY} or {@link ExchangeType#RESPONSE_STREAMING} is set.
      */
-    default ExchangeType exchangeType(RequestHeaders headers, Route route) {
+    @UnstableApi
+    default ExchangeType exchangeType(RequestHeaders headers, Routed<ServiceConfig> routed) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpService.java:72]: :+1:

File: core/src/main/java/com/linecorp/armeria/server/management/ThreadDumpService.java [hunk 2]
```diff
@@ -54,7 +55,7 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
     }
 
     @Override
-    public ExchangeType exchangeType(RequestHeaders headers, Route route) {
+    public ExchangeType exchangeType(RequestHeaders headers, Routed<ServiceConfig> routed) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/management/ThreadDumpService.java:58]: I'm not 100% sure, but can we pass `RoutingContext` instead?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/management/ThreadDumpService.java:58]: Possible if `RoutingContext.result()` is added.

File: core/src/main/java/com/linecorp/armeria/server/HttpService.java [hunk 3]
```diff
@@ -63,13 +63,13 @@ default HttpService decorate(DecoratingHttpServiceFunction function) {
...
      * Determines an {@link ExchangeType} for this {@link HttpService} from the given {@link RequestHeaders}
-     * and {@link Route}. By default, {@link ExchangeType#BIDI_STREAMING} is set.
+     * and {@link RoutingContext}. By default, {@link ExchangeType#BIDI_STREAMING} is set.
      *
      * <p>Note that an {@link HttpRequest} will be aggregated before serving the {@link HttpService} if
      * {@link ExchangeType#UNARY} or {@link ExchangeType#RESPONSE_STREAMING} is set.
      */
     @UnstableApi
-    default ExchangeType exchangeType(RequestHeaders headers, Routed<ServiceConfig> routed) {
+    default ExchangeType exchangeType(RequestHeaders headers, RoutingContext routingContext) {
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/HttpService.java:72]: Can we omit `headers` now, because `RoutingContext.headers()` will return the same one?

File: core/src/test/java/com/linecorp/armeria/internal/server/annotation/CompositeResponseConverterFunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+        };
+
+        CompositeResponseConverterFunction converter =
+                new CompositeResponseConverterFunction(ImmutableList.of(convertA, convertB, convertC));
+        Boolean responseStreaming = converter.isResponseStreaming(null, null);
+        assertThat(responseStreaming).isTrue();
+
+        converter = new CompositeResponseConverterFunction(ImmutableList.of(convertC, convertB, convertA));
+        responseStreaming = converter.isResponseStreaming(null, null);
+        assertThat(responseStreaming).isFalse();
```
Comment (trustin) [inline core/src/test/java/com/linecorp/armeria/internal/server/annotation/CompositeResponseConverterFunctionTest.java:91]: Could you add a test case where all three conversion functions return `null`?

File: core/src/main/java/com/linecorp/armeria/server/HttpService.java (hunk 3 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/HttpService.java]: Nice idea! Fixed at [`5374904` (#4177)](https://github.com/line/armeria/pull/4177/commits/537490436beec5c0d246e4ff14fe3ad426e385e5)

File: core/src/main/java/com/linecorp/armeria/server/VirtualHost.java [hunk 5]
```diff
@@ -385,15 +385,18 @@ public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx, boolea
         final Routed<ServiceConfig> routed = router.find(requireNonNull(routingCtx, "routingCtx"));
         switch (routed.routingResultType()) {
             case MATCHED:
+                maybeSetRoutingResult(routingCtx, routed);
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHost.java:388]: Oops, I found out that we need to update the Javadoc of this method because we don't return `Routed.empty()` anymore. Could you update it, please?

File: protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java [hunk 6]
```diff
@@ -142,6 +145,30 @@ public ProtobufResponseConverterFunction(Printer jsonPrinter) {
...
+    @Override
+    public Boolean isResponseStreaming(Type returnType, @Nullable MediaType produceType) {
+        final Class<?> clazz = typeToClass(unwrapAsyncType(returnType));
+        if (clazz == null) {
+            return null;
+        }
+
+        // Non-streaming types
+        if (Message.class.isAssignableFrom(clazz)) {
+            return false;
```
Comment (minwoox) [inline protobuf/src/main/java/com/linecorp/armeria/server/protobuf/ProtobufResponseConverterFunction.java:157]: I found out that there's case we fall through when the clazz is a Message. Could you check it, please?

https://github.com/line/armeria/pull/4177/files#diff-fd2e367dd8f05b6d82bf8ef3e10ece9a71205cb5e0229015c2408d8cc84e18c0R222

**Review by minwoox — COMMENTED** [review]
What a beautiful improvement! Thanks!

File: core/src/main/java/com/linecorp/armeria/server/VirtualHost.java (hunk 5 above)
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/VirtualHost.java:388]: So, if I understand this correctly, `Routed.empty()` could be returned only when `useFallbackService` is set to `false`?

**Comment by minwoox** [comment]
Oops, could you resolve the conflict?

**Comment by ikhoon** [comment]
> Oops, could you resolve the conflict?

Oooooops... what a big conflict. 😱

File: core/src/main/java/com/linecorp/armeria/server/VirtualHost.java [hunk 7]
```diff
@@ -384,21 +384,27 @@ public Routed<ServiceConfig> findServiceConfig(RoutingContext routingCtx) {
      *                           If {@code true}, the returned {@link Routed} will always be present.
      *
      * @return the {@link ServiceConfig} wrapped by a {@link Routed} if there's a match.
-     *         {@link Routed#empty()} if there's no match.
+     *         The fallback {@link ServiceConfig} wrapped by {@link Routed} if there's no match and
+     *         {@code useFallbackService} is set to {@code true}.
+     *         {@link Routed#empty()} if there's no match and {@code useFallbackService} is set to
+     *         {@code false}.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/VirtualHost.java:390]:
```suggestion
     *         The fallback {@link ServiceConfig} wrapped by {@link Routed} if there's no match and
     *         {@code useFallbackService} is {@code true}.
     *         {@link Routed#empty()} if there's no match and {@code useFallbackService} is {@code false}.
```

File: core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java [hunk 8]
```diff
@@ -37,6 +39,26 @@
 @FunctionalInterface
 public interface ResponseConverterFunction {
 
+    /**
+     * Returns whether an {@link HttpResponse} of an annotated service should be streamed.
+     * {@code null} if this converter cannot convert the {@code responseType} to an {@link HttpResponse}.
+     * This method is used for a performance optimization hint.
```
Comment (trustin) [inline core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:45]:
```suggestion
     * This method is used as a performance optimization hint.
```

**Review by minwoox — APPROVED** [review]
LGTM once others' comments are addressed. 👍

File: core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java [hunk 9]
```diff
@@ -37,6 +39,26 @@
...
+     * The non-streaming response is aggregated before being sent.
+     *
+     * @param returnType the return type of the annotated service.
+     * @param produceType the negotiated producible media type of the annotated service.
+     *                    {@code null} if the media type negotiation is not used for the service.
+     */
+    @UnstableApi
+    @Nullable
+    default Boolean isResponseStreaming(Type returnType, @Nullable MediaType produceType) {
+        return true;
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/annotation/ResponseConverterFunction.java:66]: It is a really good point.
Update the Javadoc for situations where `true` can be returned.

**Review by trustin — APPROVED** [review]
Thanks for the continuous effort on better performance 🙇

**Review by jrhee17 — APPROVED** [review]
Thanks @ikhoon 👍 🙇 👍

**Comment by minwoox** [comment]
Thanks, @ikhoon! 👍 👍 👍
